From b158aa4da181032474b3d49f9e785494e27f08b7 Mon Sep 17 00:00:00 2001
From: Peng Liu <liupeng256@huawei.com>
Date: Thu, 26 May 2022 02:06:10 +0000
Subject: [PATCH 2204/2592] [Backport] arm64: Add memmap reserve range check to
 avoid conflict

hulk inclusion
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/I59AN8

Reference: https://gitee.com/openeuler/kernel/commit/d1f92782bf23103d646b8d89baf751a897a2f9ec

--------------------------------

The user specificed memmap-reserve range may overlap in-use memory
region, and users are hard to avoid this due to KASLR. Thus, the
reduplicative memmap-reserve range should be ignored. Furthermore,
to be consistent with INITRD, the range that not in a memory region
will also be ignored.

Fixes: d05cfbd95ab2 ("arm64: Add support for memmap kernel parameters")
Signed-off-by: Peng Liu <liupeng256@huawei.com>
---
 .../admin-guide/kernel-parameters.txt         |  3 ++-
 arch/arm64/mm/init.c                          | 22 ++++++++++++++++++-
 2 files changed, 23 insertions(+), 2 deletions(-)

diff --git a/Documentation/admin-guide/kernel-parameters.txt b/Documentation/admin-guide/kernel-parameters.txt
index bb6b85b8705c..dcc991901035 100644
--- a/Documentation/admin-guide/kernel-parameters.txt
+++ b/Documentation/admin-guide/kernel-parameters.txt
@@ -2842,7 +2842,8 @@
 			[KNL,ACPI] Mark specific memory as reserved.
 			Region of memory to be reserved is from ss to ss+nn.
 			For ARM64, reserved memory must be in the range of
-			existed memory.
+			existed memory and do not overlap in-use memory region,
+			otherwise request will be ignored.
 			Example: Exclude memory from 0x18690000-0x1869ffff
 			         memmap=64K$0x18690000
 			         or
diff --git a/arch/arm64/mm/init.c b/arch/arm64/mm/init.c
index b99f818a7c21..7119dd619490 100644
--- a/arch/arm64/mm/init.c
+++ b/arch/arm64/mm/init.c
@@ -411,7 +411,27 @@ static void __init reserve_memmap_regions(void)
 	for (i = 0; i < mbk_memmap_cnt; i++) {
 		base = mbk_memmap_regions[i].base;
 		size = mbk_memmap_regions[i].size;
-		memblock_reserve(base, size);
+
+		if (!memblock_is_region_memory(base, size)) {
+			pr_warn("memmap reserve: 0x%08llx - 0x%08llx is not a memory region - ignore\n",
+				base, base + size);
+			continue;
+		}
+
+		if (memblock_is_region_reserved(base, size)) {
+			pr_warn("memmap reserve: 0x%08llx - 0x%08llx overlaps in-use memory region - ignore\n",
+				base, base + size);
+			continue;
+		}
+
+		if (memblock_reserve(base, size)) {
+			pr_warn("memmap reserve: 0x%08llx - 0x%08llx failed\n",
+				base, base + size);
+			continue;
+		}
+
+		pr_info("memmap reserved: 0x%08llx - 0x%08llx (%lld MB)",
+			base, base + size, size >> 20);
 		memblock_mark_memmap(base, size);
 	}
 }
-- 
2.17.1

